File: app/src/main/java/com/mgaetan89/showsrage/network/SickRageServices.java
package com.mgaetan89.showsrage.network;

import com.mgaetan89.showsrage.model.ApiKey;
import com.mgaetan89.showsrage.model.ComingEpisodes;
import com.mgaetan89.showsrage.model.Episodes;
import com.mgaetan89.showsrage.model.GenericResponse;
import com.mgaetan89.showsrage.model.Histories;
import com.mgaetan89.showsrage.model.LogLevel;
import com.mgaetan89.showsrage.model.Logs;
import com.mgaetan89.showsrage.model.SearchResults;
import com.mgaetan89.showsrage.model.Seasons;
import com.mgaetan89.showsrage.model.ShowStatsWrapper;
import com.mgaetan89.showsrage.model.Shows;
import com.mgaetan89.showsrage.model.ShowsStats;
import com.mgaetan89.showsrage.model.SingleEpisode;
import com.mgaetan89.showsrage.model.SingleShow;
import com.mgaetan89.showsrage.model.UpdateResponseWrapper;

import java.util.Map;

import retrofit.Callback;
import retrofit.http.GET;
import retrofit.http.Query;
import retrofit.http.QueryMap;

public interface SickRageServices {
	@GET("/{api_path}/{api_key}/?cmd=show.addnew")
	void addNewShow(@Query("indexerid") int indexerId, Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=sb.checkversion")
	void checkForUpdate(Callback<UpdateResponseWrapper> callback);

	@GET("/{api_path}/{api_key}/?cmd=history.clear")
	void clearHistory(Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=show.delete")
	void deleteShow(@Query("indexerid") int indexerId, @Query("removefiles") int removeFiles, Callback<GenericResponse> callback);

	@GET("/getkey")
	void getApiKey(@Query("u") String username, @Query("p") String password, Callback<ApiKey> callback);

	@GET("/{api_path}/{api_key}/?cmd=future")
	void getComingEpisodes(Callback<ComingEpisodes> callback);

	@GET("/{api_path}/{api_key}/?cmd=episode")
	void getEpisode(@Query("indexerid") int indexerId, @Query("season") int season, @Query("episode") int episode, Callback<SingleEpisode> callback);

	@GET("/{api_path}/{api_key}/?cmd=show.seasons")
	void getEpisodes(@Query("indexerid") int indexerId, @Query("season") int season, Callback<Episodes> callback);

	@GET("/{api_path}/{api_key}/?cmd=history")
	void getHistory(Callback<Histories> callback);

	@GET("/{api_path}/{api_key}/?cmd=logs")
	void getLogs(@Query("min_level") LogLevel minLevel, Callback<Logs> callback);

	@GET("/{api_path}/{api_key}/?cmd=show.seasonlist")
	void getSeasons(@Query("indexerid") int indexerId, Callback<Seasons> callback);

	@GET("/{api_path}/{api_key}/?cmd=show")
	void getShow(@Query("indexerid") int indexerId, Callback<SingleShow> callback);

	@GET("/{api_path}/{api_key}/?cmd=shows&sort=name")
	void getShows(Callback<Shows> callback);

	@GET("/{api_path}/{api_key}/?cmd=shows.stats")
	void getShowsStats(Callback<ShowsStats> callback);

	@GET("/{api_path}/{api_key}/")
	void getShowStats(@Query("cmd") String commands, @QueryMap Map<String, Integer> parameters, Callback<ShowStatsWrapper> callback);

	@GET("/{api_path}/{api_key}/?cmd=sb.ping")
	void ping(Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=show.pause")
	void pauseShow(@Query("indexerid") int indexerId, @Query("pause") int pause, Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=sb.restart")
	void restart(Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=sb.searchindexers")
	void search(@Query("name") String name, Callback<SearchResults> callback);

	@GET("/{api_path}/{api_key}/?cmd=episode.search")
	void searchEpisode(@Query("indexerid") int indexerId, @Query("season") int season, @Query("episode") int episode, Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=episode.setstatus")
	void setEpisodeStatus(@Query("indexerid") int indexerId, @Query("season") int season, @Query("episode") int episode, @Query("force") int force, @Query("status") String status, Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=show.setquality")
	void setShowQuality(@Query("indexerid") int indexerId, @Query("initial") String allowed, @Query("archive") String preferred, Callback<GenericResponse> callback);

	@GET("/{api_path}/{api_key}/?cmd=sb.update")
	void updateSickRage(Callback<GenericResponse> callback);
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/network/SickRageServices.kt
package com.mgaetan89.showsrage.network

import com.mgaetan89.showsrage.model.ApiKey
import com.mgaetan89.showsrage.model.Episodes
import com.mgaetan89.showsrage.model.GenericResponse
import com.mgaetan89.showsrage.model.Histories
import com.mgaetan89.showsrage.model.LogLevel
import com.mgaetan89.showsrage.model.Logs
import com.mgaetan89.showsrage.model.RootDirs
import com.mgaetan89.showsrage.model.Schedules
import com.mgaetan89.showsrage.model.SearchResults
import com.mgaetan89.showsrage.model.Seasons
import com.mgaetan89.showsrage.model.ShowStatsWrapper
import com.mgaetan89.showsrage.model.Shows
import com.mgaetan89.showsrage.model.ShowsStats
import com.mgaetan89.showsrage.model.SingleEpisode
import com.mgaetan89.showsrage.model.SingleShow
import com.mgaetan89.showsrage.model.UpdateResponseWrapper
import retrofit.Callback
import retrofit.http.GET
import retrofit.http.Query
import retrofit.http.QueryMap

interface SickRageServices {
    @GET("/{api_path}/{api_key}/?cmd=show.addnew")
    fun addNewShow(@Query("indexerid") indexerId: Int, @Query("archive") preferredQuality: String?, @Query("initial") allowedQuality: String?, @Query("status") status: String?, @Query("lang") language: String?, @Query("anime") anime: Int, @Query("subtitles") subtitles: Int, @Query("location") location: String?, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=sb.checkversion")
    fun checkForUpdate(callback: Callback<UpdateResponseWrapper>)

    @GET("/{api_path}/{api_key}/?cmd=history.clear")
    fun clearHistory(callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=show.delete")
    fun deleteShow(@Query("indexerid") indexerId: Int, @Query("removefiles") removeFiles: Int, callback: Callback<GenericResponse>)

    @GET("/{web_root}getkey")
    fun getApiKey(@Query("u") username: String?, @Query("p") password: String?, callback: Callback<ApiKey>)

    @GET("/{api_path}/{api_key}/?cmd=episode&full_path=1")
    fun getEpisode(@Query("indexerid") indexerId: Int, @Query("season") season: Int, @Query("episode") episode: Int, callback: Callback<SingleEpisode>)

    @GET("/{api_path}/{api_key}/?cmd=show.seasons")
    fun getEpisodes(@Query("indexerid") indexerId: Int?, @Query("season") season: Int, callback: Callback<Episodes>)

    @GET("/{api_path}/{api_key}/?cmd=history")
    fun getHistory(callback: Callback<Histories>)

    @GET("/{api_path}/{api_key}/?cmd=history")
    fun getHistory(): Histories

    @GET("/{api_path}/{api_key}/?cmd=logs")
    fun getLogs(@Query("min_level") minLevel: LogLevel, callback: Callback<Logs>)

    @GET("/{api_path}/{api_key}/?cmd=sb.getrootdirs")
    fun getRootDirs(callback: Callback<RootDirs>)

    @GET("/{api_path}/{api_key}/?cmd=future")
    fun getSchedule(callback: Callback<Schedules>)

    @GET("/{api_path}/{api_key}/?cmd=show.seasonlist")
    fun getSeasons(@Query("indexerid") indexerId: Int, @Query("sort") sort: String, callback: Callback<Seasons>)

    @GET("/{api_path}/{api_key}/?cmd=show")
    fun getShow(@Query("indexerid") indexerId: Int, callback: Callback<SingleShow>)

    @GET("/{api_path}/{api_key}/?cmd=shows&sort=name")
    fun getShows(callback: Callback<Shows>)

    @GET("/{api_path}/{api_key}/?cmd=shows.stats")
    fun getShowsStats(callback: Callback<ShowsStats>)

    @GET("/{api_path}/{api_key}/")
    fun getShowStats(@Query("cmd") commands: String, @QueryMap parameters: Map<String, Int>, callback: Callback<ShowStatsWrapper>)

    @GET("/{api_path}/{api_key}/?cmd=sb.ping")
    fun ping(callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=show.pause")
    fun pauseShow(@Query("indexerid") indexerId: Int, @Query("pause") pause: Int, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=postprocess&type=manual")
    fun postProcess(@Query("is_priority") replace: Int, @Query("force_replace") forceProcessing: Int, @Query("process_method") processingMethod: String?, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=show.refresh")
    fun rescanShow(@Query("indexerid") indexerId: Int, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=sb.restart")
    fun restart(callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=sb.searchindexers")
    fun search(@Query("name") name: String, callback: Callback<SearchResults>)

    @GET("/{api_path}/{api_key}/?cmd=episode.search")
    fun searchEpisode(@Query("indexerid") indexerId: Int, @Query("season") season: Int, @Query("episode") episode: Int, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=episode.subtitlesearch")
    fun searchSubtitles(@Query("indexerid") indexerId: Int, @Query("season") season: Int, @Query("episode") episode: Int, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=episode.setstatus")
    fun setEpisodeStatus(@Query("indexerid") indexerId: Int, @Query("season") season: Int, @Query("episode") episode: Int, @Query("force") force: Int, @Query("status") status: String, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=show.setquality")
    fun setShowQuality(@Query("indexerid") indexerId: Int, @Query("initial") allowed: String?, @Query("archive") preferred: String?, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=show.update")
    fun updateShow(@Query("indexerid") indexerId: Int, callback: Callback<GenericResponse>)

    @GET("/{api_path}/{api_key}/?cmd=sb.update")
    fun updateSickRage(callback: Callback<GenericResponse>)
}
